### PR TITLE
Reconciler vsphere cluster and machineconfig

### PR DESCRIPTION
**Pull request by mdsgabriel**
*Issue #, if available:*

*Description of changes:*

Validate machine config and vsphere cluster spec

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: controllers/controllers/cluster_controller.go [hunk 1]
```diff
@@ -72,6 +83,19 @@ func (r *ClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_
 		return ctrl.Result{}, nil
 	}
 
+	// Fetch the VsphereDatacenter object
+	dc := &anywherev1.VSphereDatacenterConfig{}
```
Comment (jiayiwang7) [inline controllers/controllers/cluster_controller.go:87]: nit: maybe name the var `datacenterconfig` instead of short? i did the same thing in original controller and found out later it's a bit hard to follow. (vmc, vdc, etc lol)

Comment (mdsgabriel) [reply controllers/controllers/cluster_controller.go:86]: done

File: controllers/controllers/clusters/vsphere.go [hunk 2]
```diff
@@ -2,19 +2,153 @@ package clusters
...
+	v.Log.V(4).Info("Using cp machine config %v", controlplaneMachine)
+	machineConfigMap[cluster.Spec.ControlPlaneConfiguration.MachineGroupRef.Name] = controlplaneMachine
+
+	workerNodeMachine := &anywherev1.VSphereMachineConfig{}
+	workerNodeMachineName := types.NamespacedName{Namespace: cluster.Namespace, Name: cluster.Spec.WorkerNodeGroupConfigurations[0].MachineGroupRef.Name}
+	if err := v.Client.Get(ctx, workerNodeMachineName, workerNodeMachine); err != nil {
+		return reconciler.Result{}, err
+	}
+	v.Log.V(4).Info("Using wn machine config %v", workerNodeMachine)
+	machineConfigMap[cluster.Spec.WorkerNodeGroupConfigurations[0].MachineGroupRef.Name] = workerNodeMachine
```
Comment (g-gaston) [inline controllers/controllers/clusters/vsphere.go:141]: could we do this in loop over `cluster.Spec.WorkerNodeGroupConfigurations`?
We are gonna support multiple worker node groups very soon, it would save us some time to come and refactor this

I would vote to not make the "one single worker node group" assumption in any of the new code

File: controllers/controllers/clusters/vsphere.go [hunk 3]
```diff
@@ -2,19 +2,153 @@ package clusters
...
+		Namespace: "eksa-system",
+		Name:      vsphere.CredentialsObjectName,
+	}
+	if err := v.Client.Get(ctx, secretKey, secret); err != nil {
+		return nil, err
+	}
+	return secret, nil
+}
+
+func (v *VSphereReconciler) SetupEnvsAndDefaults(ctx context.Context, vsphereDatacenter *anywherev1.VSphereDatacenterConfig) error {
```
Comment (g-gaston) [inline controllers/controllers/clusters/vsphere.go:62]: I like we are reusing this code for the multiple controllers ❤️ 

However, could we keep this as standalone methods (or maybe on a separate struct)?
It's ok for now, but I fear this might end up being another "vsphere provider" case, where if we add everything to one struct it would become to difficult to handle

File: controllers/controllers/clusters/vsphere.go [hunk 4]
```diff
@@ -2,19 +2,153 @@ package clusters
...
+}
+
+func (v *VSphereClusterReconciler) Reconcile(ctx context.Context, cluster *anywherev1.Cluster) (reconciler.Result, error) {
+	dataCenterConfig := &anywherev1.VSphereDatacenterConfig{}
+	dataCenterName := types.NamespacedName{Namespace: cluster.Namespace, Name: cluster.Spec.DatacenterRef.Name}
+	if err := v.Client.Get(ctx, dataCenterName, dataCenterConfig); err != nil {
+		return reconciler.Result{}, err
+	}
+	// Set up envs for executing Govc cmd and default values for datacenter config
+	if err := v.SetupEnvsAndDefaults(ctx, dataCenterConfig); err != nil {
```
Comment (g-gaston) [inline controllers/controllers/clusters/vsphere.go:110]: This is calling `SetDefaultsForDatacenterConfig` as well and I think we should try to avoid that in this flow
Maybe we can separate setting envars and defaults in two methods?

File: controllers/controllers/clusters/vsphere.go (hunk 2 above)
Comment (g-gaston) [inline controllers/controllers/clusters/vsphere.go:141]: Maybe we could also dry this code by building a set of machine group refs? And then iterating over it to build the map
Oh you can actually get the set for free with `cluster.MachineConfigRefs()`

File: controllers/controllers/clusters/vsphere.go [hunk 5]
```diff
@@ -2,19 +2,153 @@ package clusters
...
+	machineConfigMap[cluster.Spec.WorkerNodeGroupConfigurations[0].MachineGroupRef.Name] = workerNodeMachine
+
+	clusterSpec, err := v.FetchAppliedSpec(ctx, cluster)
+	if err != nil {
+		return reconciler.Result{}, err
+	}
+	vshepreClusterSpec := vsphere.NewSpec(clusterSpec, machineConfigMap, dataCenterConfig)
+
+	if err := v.Validator.ValidateCluster(ctx, vshepreClusterSpec); err != nil {
+		return reconciler.Result{}, err
```
Comment (g-gaston) [inline controllers/controllers/clusters/vsphere.go:150]: Should we set the failure reason, etc. in the cluster status?

File: controllers/controllers/resource/fetcher.go [hunk 6]
```diff
@@ -41,35 +41,35 @@ type ResourceFetcher interface {
...
 
-type capiResourceFetcher struct {
+type CapiResourceFetcher struct {
 	client client.Reader
 	log    logr.Logger
 }
 
-func NewCAPIResourceFetcher(client client.Reader, Log logr.Logger) *capiResourceFetcher {
-	return &capiResourceFetcher{
+func NewCAPIResourceFetcher(client client.Reader, Log logr.Logger) *CapiResourceFetcher {
```
Comment (jiayiwang7) [inline controllers/controllers/resource/fetcher.go:49]: are we still using the old controller fetcher.go? somehow i thought we are moving away from it

File: controllers/controllers/clusters/vsphere.go [hunk 7]
```diff
@@ -2,19 +2,153 @@ package clusters
...
+}
+
+func (v *VSphereClusterReconciler) FetchAppliedSpec(ctx context.Context, cs *anywherev1.Cluster) (*c.Spec, error) {
+	return c.BuildSpecForCluster(ctx, cs, v.bundles, nil)
+}
+
+func (v *VSphereClusterReconciler) Reconcile(ctx context.Context, cluster *anywherev1.Cluster) (reconciler.Result, error) {
+	dataCenterConfig := &anywherev1.VSphereDatacenterConfig{}
+	dataCenterName := types.NamespacedName{Namespace: cluster.Namespace, Name: cluster.Spec.DatacenterRef.Name}
+	if err := v.Client.Get(ctx, dataCenterName, dataCenterConfig); err != nil {
```
Comment (jiayiwang7) [inline controllers/controllers/clusters/vsphere.go:106]: wonder if we should pass down the datacenterconfig fetched from https://github.com/aws/eks-anywhere/pull/921/files#diff-58f6d1beb0dffe9177008c7ea4959b2c4a4543ca561e8fe7b6cdf98114595272R86
instead of get it again using client

File: controllers/controllers/clusters/vsphere.go [hunk 8]
```diff
@@ -2,19 +2,153 @@ package clusters
...
+	}
+	v.Log.V(4).Info("Using cp machine config %v", controlplaneMachine)
+	machineConfigMap[cluster.Spec.ControlPlaneConfiguration.MachineGroupRef.Name] = controlplaneMachine
+
+	workerNodeMachine := &anywherev1.VSphereMachineConfig{}
+	workerNodeMachineName := types.NamespacedName{Namespace: cluster.Namespace, Name: cluster.Spec.WorkerNodeGroupConfigurations[0].MachineGroupRef.Name}
+	if err := v.Client.Get(ctx, workerNodeMachineName, workerNodeMachine); err != nil {
+		return reconciler.Result{}, err
+	}
+	v.Log.V(4).Info("Using wn machine config %v", workerNodeMachine)
```
Comment (jiayiwang7) [inline controllers/controllers/clusters/vsphere.go:140]: nit: maybe use "worker node" instead of "wn"? :) same for "cp" above

File: controllers/controllers/clusters/vsphere.go [hunk 9]
```diff
@@ -2,19 +2,153 @@ package clusters
...
+		v.Log.Error(err, "Failed to set up env vars and default values for VsphereDatacenterConfig")
+		return reconciler.Result{}, err
+	}
+
+	machineConfigMap := map[string]*anywherev1.VSphereMachineConfig{}
+
+	if cluster.Spec.ExternalEtcdConfiguration != nil {
+		mc := &anywherev1.VSphereMachineConfig{}
+		mcName := types.NamespacedName{Namespace: cluster.Namespace, Name: cluster.Spec.ExternalEtcdConfiguration.MachineGroupRef.Name}
+		if err := v.Client.Get(ctx, mcName, mc); err != nil {
```
Comment (jiayiwang7) [inline controllers/controllers/clusters/vsphere.go:120]: I see some of the objects are fetched by method defined in fetcher.go. Some (like this one) are directly calling the `Client.Get`. would be nice if we can stay consistent and use one way over another

File: controllers/controllers/cluster_controller.go [hunk 10]
```diff
@@ -72,6 +82,19 @@ func (r *ClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_
...
+	dataCenterName := types.NamespacedName{Namespace: cluster.Namespace, Name: cluster.Spec.DatacenterRef.Name}
+	if err := r.client.Get(ctx, dataCenterName, dataCenterConfig); err != nil {
+		return ctrl.Result{}, err
+	}
+	log.Info("Using data center config config", "datacenter", dataCenterConfig)
+
+	if !dataCenterConfig.Status.SpecValid {
+		log.Info("Skipping cluster reconciliation because data center config is invalid %v", dataCenterConfig)
+		return ctrl.Result{}, nil
+	}
```
[the remark this answers is not in the document]
Comment (mdsgabriel) [reply controllers/controllers/cluster_controller.go]: done

File: controllers/controllers/cluster_controller.go [hunk 11]
```diff
@@ -37,7 +48,6 @@ func (r *ClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
 // TODO: add here kubebuilder permissions as neeeded
```
[the remark this answers is not in the document]
Comment (mdsgabriel) [reply controllers/controllers/cluster_controller.go]: done. copied the permissions from the old controller

File: pkg/providers/vsphere/validator.go [hunk 12]
```diff
@@ -62,7 +62,7 @@ func (v *Validator) ValidateVCenterConfig(ctx context.Context, datacenterConfig
 }
 
 // TODO: dry out machine configs validations
-func (v *Validator) validateCluster(ctx context.Context, vsphereClusterSpec *spec) error {
+func (v *Validator) ValidateCluster(ctx context.Context, vsphereClusterSpec *Spec) error {
```
[the remark this answers is not in the document]
Comment (mdsgabriel) [reply pkg/providers/vsphere/validator.go]: done

File: controllers/controllers/clusters/vsphere.go (hunk 2 above)
Comment (mdsgabriel) [reply controllers/controllers/clusters/vsphere.go]: done

Comment (mdsgabriel) [reply controllers/controllers/clusters/vsphere.go]: done

File: controllers/controllers/clusters/vsphere.go (hunk 7 above)
Comment (mdsgabriel) [reply controllers/controllers/clusters/vsphere.go:106]: The previous call has been removed

File: controllers/controllers/clusters/vsphere.go (hunk 8 above)
Comment (mdsgabriel) [reply controllers/controllers/clusters/vsphere.go]: removed

File: controllers/controllers/clusters/vsphere.go [hunk 13]
```diff
@@ -2,19 +2,153 @@ package clusters
...
+	VSphereReconciler
 	*providerClusterReconciler
+
+	capiResourceFetcher *resource.CapiResourceFetcher
 }
 
-func NewVSphereReconciler() *VSphereReconciler {
-	return &VSphereReconciler{providerClusterReconciler: &providerClusterReconciler{}}
+func NewVSphereReconciler(client client.Client, log logr.Logger, validator *vsphere.Validator, defaulter *vsphere.Defaulter) *VSphereClusterReconciler {
+	capiResourceFetcher := resource.NewCAPIResourceFetcher(client, log)
```
[the remark this answers is not in the document]
Comment (mdsgabriel) [reply controllers/controllers/clusters/vsphere.go]: done

File: controllers/controllers/clusters/vsphere.go (hunk 9 above)
Comment (mdsgabriel) [reply controllers/controllers/clusters/vsphere.go]: the resource fetcher has been removed. The current code is using Client.Get

File: controllers/controllers/clusters/vsphere.go (hunk 4 above)
Comment (mdsgabriel) [reply controllers/controllers/clusters/vsphere.go]: done

File: controllers/controllers/resource/fetcher.go (hunk 6 above)
Comment (mdsgabriel) [reply controllers/controllers/resource/fetcher.go:49]: not anymore. The code is now using Client.Get

File: controllers/controllers/clusters/vsphere.go (hunk 3 above)
Comment (mdsgabriel) [reply controllers/controllers/clusters/vsphere.go]: done

**Review by jiayiwang7 — APPROVED** [review]

File: controllers/controllers/cluster_controller_test.go [hunk 14]
```diff
@@ -0,0 +1,378 @@
...
+	machineConfigCP := createCPMachineConfig()
+	machineConfigWN := createWNMachineConfig()
+
+	objs := []runtime.Object{cluster, datacenterConfig, secret, bundle, machineConfigCP, machineConfigWN}
+
+	s := scheme.Scheme
+	s.AddKnownTypes(anywherev1.GroupVersion, cluster)
+	s.AddKnownTypes(anywherev1.GroupVersion, datacenterConfig)
+	s.AddKnownTypes(anywherev1.GroupVersion, bundle)
+	s.AddKnownTypes(anywherev1.GroupVersion, machineConfigCP)
```
Comment (g-gaston) [inline controllers/controllers/cluster_controller_test.go:47]: nit: we can avoid this by `import _ "github.com/aws/eks-anywhere/internal/test/envtest"`

File: controllers/controllers/cluster_controller_test.go [hunk 15]
```diff
@@ -0,0 +1,378 @@
...
+	}
+
+	apiCluster := &anywherev1.Cluster{}
+
+	err = r.client.Get(context.TODO(), req.NamespacedName, apiCluster)
+	if err != nil {
+		t.Fatalf("get cluster: (%v)", err)
+	}
+	if apiCluster.Status.FailureMessage != nil {
+		t.Errorf("Expected failure message to be nil")
```
Comment (g-gaston) [inline controllers/controllers/cluster_controller_test.go:87]: nit: include the failure reason in the error message?

**Review by g-gaston — APPROVED** [review]
Awesome!
Let me know when you wanna merge

File: controllers/controllers/cluster_controller_test.go (hunk 15 above)
Comment (mdsgabriel) [reply controllers/controllers/cluster_controller_test.go]: done

File: controllers/controllers/cluster_controller_test.go (hunk 14 above)
Comment (mdsgabriel) [reply controllers/controllers/cluster_controller_test.go]: done

**Review by g-gaston — COMMENTED** [review]
/lgtm
